src/components/CalendarContainer.ts

```diff
@@ -0,0 +1,293 @@
+        };
+    }
+
+    render() {
+        if (this.state.alertMessage) {
+            return createElement(Alert, {
+                bootstrapStyle: "danger",
+                className: "widget-calendar-alert",
+                message: this.state.alertMessage
+            });
+        } else {
+            return createElement(Calendar, {
+                events: this.state.events,
+                defaultView: this.props.defaultView,
+                popup: this.props.popup,
+                selectable: this.props.selectable,
+                onSelectEventAction: this.onClickEvent,
+                onSelectSlotAction: this.onClickSlot,
+                onEventDropAction: this.onDropEvent
+            });
+        }
+    }
+
+    componentDidMount() {
+        if (!this.state.alertMessage && this.props.mxObject) {
+            this.fetchData(this.props.mxObject);
+        }
+    }
+
+    componentWillUnMount() {
+        this.subscriptionHandles.forEach(window.mx.data.unsubscribe);
+    }
+
+    componentWillReceiveProps(nextProps: CalendarContainerProps) {
+        this.fetchData(nextProps.mxObject);
+        this.resetSubscriptions(nextProps.mxObject);
+    }
+
+    private fetchData = (mxObject: mendix.lib.MxObject) => {
+        const { dataSource, dataSourceMicroflow, eventEntity } = this.props;
+        if (dataSource === "XPath" && eventEntity && mxObject) {
+            this.fetchEventsByXPath(mxObject.getGuid());
+        } else if (dataSource === "microflow" && dataSourceMicroflow) {
+            this.fetchEventsByMicroflow(mxObject);
+        }
+    }
+
+    private resetSubscriptions = (mxObject: mendix.lib.MxObject) => {
+        this.subscriptionHandles.forEach(window.mx.data.unsubscribe);
+        this.subscriptionHandles = [];
+
+        if (mxObject) {
+            this.subscriptionHandles.push(window.mx.data.subscribe({
+                callback: () => this.fetchData(mxObject),
+                entity: this.props.eventEntity
+            }));
+            this.subscriptionHandles.push(window.mx.data.subscribe({
+                callback: () => this.fetchData(mxObject),
+                guid: mxObject.getGuid()
+            }));
+            [
+                this.props.titleAttribute,
+                this.props.startAttribute,
+                this.props.endAttribute
+            ].forEach(attr => this.subscriptionHandles.push(window.mx.data.subscribe({
+                attr,
+                callback: () => this.fetchData(mxObject), guid: mxObject.getGuid()
+            })));
+        }
+    }
+
+    private fetchEventsByXPath(contextGuid: string) {
+        const { entityConstraint, eventEntity } = this.props;
+        const constraint = entityConstraint ? entityConstraint.replace(/\[%CurrentObject%\]/g, contextGuid) : "";
+        const XPath = `//${eventEntity}${constraint}`;
+
+        window.mx.data.get({
+            xpath: XPath,
+            callback: mxObjects => this.setEventsFromMxObjects(mxObjects)
```

Comment:
Already gave u feedback on this, but it's been reverted for some reason...

Reply:
Had'nt pushed my code